tikv/raftstore/snap_builder.go

```diff
@@ -0,0 +1,231 @@
+package raftstore
+
+import (
+	"bytes"
+	"os"
+
+	"github.com/coocood/badger"
+	"github.com/ngaut/unistore/lockstore"
+	"github.com/ngaut/unistore/rocksdb"
+	"github.com/ngaut/unistore/tikv/mvcc"
+	"github.com/pingcap/errors"
+	"github.com/pingcap/kvproto/pkg/kvrpcpb"
+	"github.com/pingcap/kvproto/pkg/metapb"
+	"github.com/pingcap/tidb/util/codec"
+	"math"
+)
+
+func newSnapBuilder(cfFiles []*CFFile, db *badger.DB, lockStore *lockstore.MemStore, region *metapb.Region) (*snapBuilder, error) {
+	b := new(snapBuilder)
+	b.cfFiles = cfFiles
+	b.endKey = rawRegionKey(region.EndKey)
+	b.txn = db.NewTransaction(false)
+	b.dbIterator = b.txn.NewIterator(badger.DefaultIteratorOptions)
+	startKey := rawRegionKey(region.StartKey)
+	b.dbIterator.Seek(startKey)
+	if b.dbIterator.Valid() && !b.reachEnd(b.dbIterator.Item().Key()) {
+		b.curDBKey = b.dbIterator.Item().Key()
+	}
+	b.dbOldIterator = b.txn.NewIterator(badger.DefaultIteratorOptions)
+	b.dbOldIterator.Seek(mvcc.EncodeOldKey(startKey, math.MaxUint64))
+
+	b.lockIterator = lockStore.NewIterator()
+	b.lockIterator.Seek(startKey)
+	if b.lockIterator.Valid() && !b.reachEnd(b.lockIterator.Key()) {
+		b.curLockKey = b.lockIterator.Key()
+	}
+	lockCFFile := cfFiles[lockCFIdx].File
+	if lockCFFile == nil {
+		return nil, errors.New("lock CF file is nil")
+	}
+	b.lockCFWriter = cfFiles[lockCFIdx].File
+	b.defaultCFWriter = cfFiles[defaultCFIdx].SstWriter
```

Comment:
Also need to verify default cf sstWriter and write cf sstWriter like lock cf file?